File: tradebot/storage/repo.py
from __future__ import annotations

from dataclasses import asdict, dataclass, field
from datetime import date as _date
from datetime import datetime
from pathlib import Path
from typing import Any, Literal

import pandas as pd

from tradebot.storage.atomic import atomic_write_json, read_json_or_default

Mode = Literal["demo", "real", "backtest"]
Side = Literal["buy", "sell"]


@dataclass
class Trade:
    mode: Mode
    pair: str
    side: Side
    base_amount: float
    quote_amount: float
    price: float
    fee_quote: float
    slippage_pct: float
    tx_signature: str | None
    opened_at: datetime
    confidence: float | None
    notes: str | None


@dataclass
class PositionRecord:
    pair: str
    base_amount: float
    avg_entry_price: float
    fees_paid_quote: float = 0.0


@dataclass
class PortfolioState:
    mode: Mode
    cash: float
    realized_pnl_total: float
    equity_high: float
    positions: list[PositionRecord] = field(default_factory=list)
    sol_balance: float = 0.0
    sol_gas_paid_total: float = 0.0


@dataclass
class OHLCVCandle:
    pair: str
    timeframe: str
    bucket_start: datetime
    open: float
    high: float
    low: float
    close: float
    volume_quote: float = 0.0


def _safe_pair(pair: str) -> str:
    return pair.replace("/", "--")


class JsonStorage:
    def __init__(self, root: Path) -> None:
        self._root = Path(root)
        self._root.mkdir(parents=True, exist_ok=True)

    # --- trades ---
    def _trades_path(self) -> Path:
        return self._root / "trades.json"

    def append_trade(self, trade: Trade) -> None:
        path = self._trades_path()
        existing = read_json_or_default(path, default=[])
        d = asdict(trade)
        d["opened_at"] = trade.opened_at.isoformat()
        existing.append(d)
        atomic_write_json(path, existing)

    def round_trip_returns(
        self, mode: Mode, pair: str | None = None, limit: int = 200
    ) -> list[float]:
        """Compute per-trade round-trip returns from trade history.

        Walks trades chronologically (storage is newest-first; we reverse).
        For each buy→sell pair:
            return = (sell_quote_net - buy_quote_cost) / buy_quote_cost
        where buy_quote_cost = quote_amount + fee_quote
        and   sell_quote_net = quote_amount - fee_quote.
        Partial sells are treated as full closes, documented approximation.
        """
        trades = self.list_trades(mode=mode, limit=limit * 2)
        if pair:
            trades = [t for t in trades if t.pair == pair]
        trades.reverse()  # chronological order
        returns: list[float] = []
        last_buy_quote: float | None = None
        last_buy_base: float | None = None
        for t in trades:
            if t.side == "buy":
                last_buy_quote = t.quote_amount + t.fee_quote
                last_buy_base = t.base_amount
            elif t.side == "sell" and last_buy_quote is not None and last_buy_base is not None:
                sell_quote = t.quote_amount - t.fee_quote
                ret = (sell_quote - last_buy_quote) / last_buy_quote
                returns.append(ret)
                last_buy_quote = None
                last_buy_base = None
        return returns

    def list_trades(self, mode: Mode, limit: int = 50) -> list[Trade]:
        rows = read_json_or_default(self._trades_path(), default=[])
        filtered = [r for r in rows if r.get("mode") == mode]
        filtered.reverse()  # newest first
        out: list[Trade] = []
        for r in filtered[:limit]:
            out.append(
                Trade(
                    mode=r["mode"],
                    pair=r["pair"],
                    side=r["side"],
                    base_amount=r["base_amount"],
                    quote_amount=r["quote_amount"],
                    price=r["price"],
                    fee_quote=r["fee_quote"],
                    slippage_pct=r["slippage_pct"],
                    tx_signature=r.get("tx_signature"),
                    opened_at=datetime.fromisoformat(r["opened_at"]),
                    confidence=r.get("confidence"),
                    notes=r.get("notes"),
                )
            )
        return out

    # --- portfolio state ---
    def _portfolio_path(self, mode: Mode) -> Path:
        return self._root / f"portfolio.{mode}.json"

    def save_portfolio_state(self, state: PortfolioState) -> None:
        d = {
            "mode": state.mode,
            "cash": state.cash,
            "realized_pnl_total": state.realized_pnl_total,
            "equity_high": state.equity_high,
            "sol_balance": state.sol_balance,
            "sol_gas_paid_total": state.sol_gas_paid_total,
            "positions": [asdict(p) for p in state.positions],
        }
        atomic_write_json(self._portfolio_path(state.mode), d)

    def load_portfolio_state(self, mode: Mode) -> PortfolioState | None:
        d = read_json_or_default(self._portfolio_path(mode), default=None)
        if d is None:
            return None
        return PortfolioState(
            mode=d["mode"],
            cash=d["cash"],
            realized_pnl_total=d["realized_pnl_total"],
            equity_high=d["equity_high"],
            sol_balance=d.get("sol_balance", 0.0),
            sol_gas_paid_total=d.get("sol_gas_paid_total", 0.0),
            positions=[PositionRecord(**p) for p in d.get("positions", [])],
        )

    # --- equity snapshots ---
    def _equity_path(self, mode: Mode) -> Path:
        return self._root / f"equity.{mode}.json"

    def append_equity_snapshot(
        self,
        mode: Mode,
        snapshot_at: datetime,
        equity: float,
        cash: float,
        positions_value: float,
    ) -> None:
        path = self._equity_path(mode)
        existing = read_json_or_default(path, default=[])
        existing.append(
            {
                "snapshot_at": snapshot_at.isoformat(),
                "equity": equity,
                "cash": cash,
                "positions_value": positions_value,
            }
        )
        atomic_write_json(path, existing)

    def list_equity_snapshots(self, mode: Mode, limit: int = 200) -> list[dict[str, Any]]:
        rows: list[dict[str, Any]] = read_json_or_default(self._equity_path(mode), default=[])
        # chronological (oldest first), tail last `limit`
        if limit and len(rows) > limit:
            rows = rows[-limit:]
        return rows

    # --- ohlcv ---
    def _ohlcv_path(self, pair: str, timeframe: str) -> Path:
        return self._root / "ohlcv" / f"{_safe_pair(pair)}__{timeframe}.json"

    def upsert_ohlcv(self, candle: OHLCVCandle) -> None:
        path = self._ohlcv_path(candle.pair, candle.timeframe)
        existing = read_json_or_default(path, default=[])
        bucket_iso = candle.bucket_start.isoformat()
        # Replace if same bucket exists (last one, append-mostly assumption)
        idx = None
        for i in range(len(existing) - 1, -1, -1):
            if existing[i]["bucket_start"] == bucket_iso:
                idx = i
                break
        record = {
            "bucket_start": bucket_iso,
            "open": candle.open,
            "high": candle.high,
            "low": candle.low,
            "close": candle.close,
            "volume_quote": candle.volume_quote,
        }
        if idx is not None:
            existing[idx] = record
        else:
            existing.append(record)
        atomic_write_json(path, existing)

    def load_ohlcv(self, pair: str, timeframe: str, limit: int = 200) -> pd.DataFrame:
        rows = read_json_or_default(self._ohlcv_path(pair, timeframe), default=[])
        if limit and len(rows) > limit:
            rows = rows[-limit:]
        if not rows:
            return pd.DataFrame(columns=["timestamp", "open", "high", "low", "close", "volume"])
        return pd.DataFrame(
            {
                "timestamp": [datetime.fromisoformat(r["bucket_start"]) for r in rows],
                "open": [r["open"] for r in rows],
                "high": [r["high"] for r in rows],
                "low": [r["low"] for r in rows],
                "close": [r["close"] for r in rows],
                "volume": [r["volume_quote"] for r in rows],
            }
        )

    def load_ohlcv_for_pairs(
        self,
        pairs: list[str],
        timeframes: list[str],
        limit: int = 200,
    ) -> dict[str, dict[str, pd.DataFrame]]:
        out: dict[str, dict[str, pd.DataFrame]] = {}
        for pair in pairs:
            out[pair] = {}
            for tf in timeframes:
                out[pair][tf] = self.load_ohlcv(pair=pair, timeframe=tf, limit=limit)
        return out

    # --- risk state ---
    def _risk_path(self, mode: Mode) -> Path:
        return self._root / f"risk_state.{mode}.json"

    def save_risk_state(self, mode: Mode, state: Any) -> None:
        d = {
            "trades_per_day": {k.isoformat(): v for k, v in state.trades_per_day.items()},
            "daily_start_equity": {k.isoformat(): v for k, v in state.daily_start_equity.items()},
            "weekly_start_equity": {k.isoformat(): v for k, v in state.weekly_start_equity.items()},
            "day_paused_until": (
                state.day_paused_until.isoformat() if state.day_paused_until else None
            ),
            "week_paused_until": (
                state.week_paused_until.isoformat() if state.week_paused_until else None
            ),
            "kill_switch_active": state.kill_switch_active,
            "kill_switch_reason": state.kill_switch_reason,
        }
        atomic_write_json(self._risk_path(mode), d)

    def load_risk_state(self, mode: Mode) -> Any | None:
        from tradebot.core.risk import RiskState  # local import to avoid cycle

        d = read_json_or_default(self._risk_path(mode), default=None)
        if d is None:
            return None
        return RiskState(
            trades_per_day={
                _date.fromisoformat(k): v for k, v in d.get("trades_per_day", {}).items()
            },
            daily_start_equity={
                _date.fromisoformat(k): v for k, v in d.get("daily_start_equity", {}).items()
            },
            weekly_start_equity={
                _date.fromisoformat(k): v for k, v in d.get("weekly_start_equity", {}).items()
            },
            day_paused_until=(
                _date.fromisoformat(d["day_paused_until"]) if d.get("day_paused_until") else None
            ),
            week_paused_until=(
                _date.fromisoformat(d["week_paused_until"]) if d.get("week_paused_until") else None
            ),
            kill_switch_active=bool(d.get("kill_switch_active", False)),
            kill_switch_reason=d.get("kill_switch_reason", ""),
        )

    # --- mark history (rolling chart cache) ---
    def _mark_history_path(self, mode: Mode) -> Path:
        return self._root / f"mark_history.{mode}.json"

    def save_mark_history(
        self,
        mode: Mode,
        history: dict[str, list[tuple[datetime, float]]],
    ) -> None:
        d = {
            pair: [{"t": ts.isoformat(), "p": float(price)} for ts, price in points]
            for pair, points in history.items()
        }
        atomic_write_json(self._mark_history_path(mode), d)

    def load_mark_history(
        self,
        mode: Mode,
    ) -> dict[str, list[tuple[datetime, float]]]:
        d = read_json_or_default(self._mark_history_path(mode), default={})
        out: dict[str, list[tuple[datetime, float]]] = {}
        for pair, points in d.items():
            out[pair] = [(datetime.fromisoformat(p["t"]), float(p["p"])) for p in points]
        return out

File: crates/storage/src/repo.rs
//! `JsonStorage`: file layout and methods mirroring `tradebot/storage/repo.py`.

use crate::atomic::{atomic_write_json, read_json_or_default};
use crate::error::StorageError;
use crate::records::{
    Candle, EquitySnapshot, MarkHistory, MarkPoint, OhlcvCandle, OhlcvRecord, PortfolioState,
    RiskStateRecord, Trade,
};
use rust_decimal::prelude::ToPrimitive;
use std::collections::BTreeMap;
use std::path::{Path, PathBuf};
use tradebot_common::Mode;

/// Replace `/` with `--` so a trading pair can be used as a filename
/// component. Mirrors `_safe_pair` in `repo.py`.
fn safe_pair(pair: &str) -> String {
    pair.replace('/', "--")
}

pub struct JsonStorage {
    root: PathBuf,
}

impl JsonStorage {
    pub fn new(root: impl Into<PathBuf>) -> Result<Self, StorageError> {
        let root = root.into();
        std::fs::create_dir_all(&root).map_err(|source| StorageError::Io {
            path: root.clone(),
            source,
        })?;
        Ok(Self { root })
    }

    pub fn root(&self) -> &Path {
        &self.root
    }

    // --- trades ---

    fn trades_path(&self) -> PathBuf {
        self.root.join("trades.json")
    }

    pub fn append_trade(&self, trade: &Trade) -> Result<(), StorageError> {
        let path = self.trades_path();
        let mut existing: Vec<Trade> = read_json_or_default(&path, Vec::new());
        existing.push(trade.clone());
        atomic_write_json(&path, &existing)
    }

    /// Trades for `mode`, newest first, capped at `limit`.
    pub fn list_trades(&self, mode: Mode, limit: usize) -> Vec<Trade> {
        let rows: Vec<Trade> = read_json_or_default(&self.trades_path(), Vec::new());
        let mut filtered: Vec<Trade> = rows.into_iter().filter(|t| t.mode == mode).collect();
        filtered.reverse();
        filtered.truncate(limit);
        filtered
    }

    /// Per-trade round-trip returns from trade history. Walks trades
    /// chronologically (storage is newest-first; we reverse). For each
    /// buy -> sell pair: `return = (sell_quote_net - buy_quote_cost) /
    /// buy_quote_cost` where `buy_quote_cost = quote_amount + fee_quote` and
    /// `sell_quote_net = quote_amount - fee_quote`. Partial sells are treated
    /// as full closes; this is a documented approximation, same as Python.
    pub fn round_trip_returns(&self, mode: Mode, pair: Option<&str>, limit: usize) -> Vec<f64> {
        let mut trades = self.list_trades(mode, limit.saturating_mul(2));
        if let Some(pair) = pair {
            trades.retain(|t| t.pair == pair);
        }
        trades.reverse(); // chronological order

        let mut returns = Vec::new();
        let mut last_buy_quote = None;
        for t in trades {
            match t.side {
                crate::records::Side::Buy => {
                    last_buy_quote = Some(t.quote_amount + t.fee_quote);
                }
                crate::records::Side::Sell => {
                    if let Some(buy_quote_cost) = last_buy_quote {
                        let sell_quote_net = t.quote_amount - t.fee_quote;
                        let ret = (sell_quote_net - buy_quote_cost) / buy_quote_cost;
                        returns.push(ret.to_f64().unwrap_or(0.0));
                        last_buy_quote = None;
                    }
                }
            }
        }
        returns
    }

    // --- portfolio state ---

    fn portfolio_path(&self, mode: Mode) -> PathBuf {
        self.root.join(format!("portfolio.{mode}.json"))
    }

    pub fn save_portfolio_state(&self, state: &PortfolioState) -> Result<(), StorageError> {
        atomic_write_json(&self.portfolio_path(state.mode), state)
    }

    pub fn load_portfolio_state(&self, mode: Mode) -> Option<PortfolioState> {
        read_json_or_default(&self.portfolio_path(mode), None)
    }

    // --- equity snapshots ---

    fn equity_path(&self, mode: Mode) -> PathBuf {
        self.root.join(format!("equity.{mode}.json"))
    }

    pub fn append_equity_snapshot(
        &self,
        mode: Mode,
        snapshot: &EquitySnapshot,
    ) -> Result<(), StorageError> {
        let path = self.equity_path(mode);
        let mut existing: Vec<EquitySnapshot> = read_json_or_default(&path, Vec::new());
        existing.push(snapshot.clone());
        atomic_write_json(&path, &existing)
    }

    /// Equity snapshots for `mode`, chronological (oldest first), tail last
    /// `limit`.
    pub fn list_equity_snapshots(&self, mode: Mode, limit: usize) -> Vec<EquitySnapshot> {
        let mut rows: Vec<EquitySnapshot> =
            read_json_or_default(&self.equity_path(mode), Vec::new());
        if limit > 0 && rows.len() > limit {
            rows = rows.split_off(rows.len() - limit);
        }
        rows
    }

    // --- ohlcv ---

    fn ohlcv_path(&self, pair: &str, timeframe: &str) -> PathBuf {
        self.root
            .join("ohlcv")
            .join(format!("{}__{timeframe}.json", safe_pair(pair)))
    }

    pub fn upsert_ohlcv(&self, candle: &OhlcvCandle) -> Result<(), StorageError> {
        let path = self.ohlcv_path(&candle.pair, &candle.timeframe);
        let mut existing: Vec<OhlcvRecord> = read_json_or_default(&path, Vec::new());

        let record = OhlcvRecord {
            bucket_start: candle.bucket_start,
            open: candle.open,
            high: candle.high,
            low: candle.low,
            close: candle.close,
            volume_quote: candle.volume_quote,
        };

        // Replace if the same bucket exists (search from the end, matching
        // the append-mostly assumption in repo.py).
        let idx = existing
            .iter()
            .rposition(|r| r.bucket_start == candle.bucket_start);
        match idx {
            Some(i) => existing[i] = record,
            None => existing.push(record),
        }
        atomic_write_json(&path, &existing)
    }

    /// The last `limit` candles for `pair`/`timeframe`, oldest first.
    pub fn load_ohlcv(&self, pair: &str, timeframe: &str, limit: usize) -> Vec<Candle> {
        let mut rows: Vec<OhlcvRecord> =
            read_json_or_default(&self.ohlcv_path(pair, timeframe), Vec::new());
        if limit > 0 && rows.len() > limit {
            rows = rows.split_off(rows.len() - limit);
        }
        rows.into_iter()
            .map(|r| Candle {
                timestamp: r.bucket_start,
                open: r.open,
                high: r.high,
                low: r.low,
                close: r.close,
                volume: r.volume_quote,
            })
            .collect()
    }

    pub fn load_ohlcv_for_pairs(
        &self,
        pairs: &[String],
        timeframes: &[String],
        limit: usize,
    ) -> BTreeMap<String, BTreeMap<String, Vec<Candle>>> {
        let mut out = BTreeMap::new();
        for pair in pairs {
            let mut per_tf = BTreeMap::new();
            for tf in timeframes {
                per_tf.insert(tf.clone(), self.load_ohlcv(pair, tf, limit));
            }
            out.insert(pair.clone(), per_tf);
        }
        out
    }

    // --- risk state ---

    fn risk_path(&self, mode: Mode) -> PathBuf {
        self.root.join(format!("risk_state.{mode}.json"))
    }

    pub fn save_risk_state(&self, mode: Mode, state: &RiskStateRecord) -> Result<(), StorageError> {
        atomic_write_json(&self.risk_path(mode), state)
    }

    pub fn load_risk_state(&self, mode: Mode) -> Option<RiskStateRecord> {
        read_json_or_default(&self.risk_path(mode), None)
    }

    // --- mark history (rolling chart cache) ---

    fn mark_history_path(&self, mode: Mode) -> PathBuf {
        self.root.join(format!("mark_history.{mode}.json"))
    }

    pub fn save_mark_history(&self, mode: Mode, history: &MarkHistory) -> Result<(), StorageError> {
        atomic_write_json(&self.mark_history_path(mode), history)
    }

    pub fn load_mark_history(&self, mode: Mode) -> MarkHistory {
        read_json_or_default(&self.mark_history_path(mode), MarkHistory::new())
    }
}

/// Convenience constructor for a single mark point, used by callers building
/// up a `MarkHistory` to pass to `save_mark_history`.
impl MarkPoint {
    pub fn new(t: chrono::DateTime<chrono::Utc>, p: tradebot_common::Money) -> Self {
        Self { t, p }
    }
}
